typing/includecore.ml

```diff
@@ -507,14 +623,28 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
       | Some err -> Some (Private_object(ty1, ty2, err))
     end
   | _ -> begin
-    match
-      match priv2 with
-      | Private -> Ctype.equal_private env params1 ty1 params2 ty2
-      | Public -> Ctype.equal env true (params1 @ [ty1]) (params2 @ [ty2])
-    with
-    | exception Ctype.Equality trace -> Some (Manifest (env, trace))
-    | () -> None
-  end
+      let is_private_abbrev_2 =
+        match priv2, kind2 with
+        | Private, Type_abstract -> begin
+            (* Same checks as the [when] guards from above, inverted *)
+            match ty2'.desc with
+            | Tvariant row ->
+                not (is_absrow env (Btype.row_more row))
+            | Tobject (fi, _) ->
+                not (is_absrow env (snd (Ctype.flatten_fields fi)))
+            | _ -> true
+          end
+        | _, _ -> false
+      in
```

Comment:
This is a bug fix. I'm not 100% sure I can hit it with actual code, but it should still be fixed. `equal_private` should only be used when checking inclusion for private type abbreviations, but was previously being used for private datatypes and private row types too.

It was included in the PR because the PR also distinguishes private type abbreviations form private datatypes and private row types in error messages. I thought we could share the detection code between here and there, so suggested that Antal fix it whilst he was there. In the end I'm not sure any actual code is shared between the two places -- so it could perhaps be its own PR instead.

Reply:
From my point of view, a separate commit could be a good compromise in order to record the information that this is a bug fix in the git history without adding too much administrative complexity.

Reply:
I've added one at the beginning of the history for this PR.